Lib/pathlib/_abc.py

```diff
@@ -848,26 +852,27 @@ def on_error(err):
                 raise err
         stack = [(self, target)]
         while stack:
-            source_dir, target_dir = stack.pop()
+            src, dst = stack.pop()
             try:
-                sources = source_dir.iterdir()
-                target_dir.mkdir(exist_ok=dirs_exist_ok)
-                if preserve_metadata:
-                    source_dir._copy_metadata(target_dir)
-                for source in sources:
-                    if ignore and ignore(source):
-                        continue
-                    try:
-                        if source.is_dir(follow_symlinks=follow_symlinks):
-                            stack.append((source, target_dir.joinpath(source.name)))
-                        else:
-                            source.copy(target_dir.joinpath(source.name),
-                                        follow_symlinks=follow_symlinks,
-                                        preserve_metadata=preserve_metadata)
-                    except OSError as err:
-                        on_error(err)
+                if not follow_symlinks and src.is_symlink():
+                    dst._symlink_to_target_of(src)
+                    if preserve_metadata:
+                        src._copy_metadata(dst, follow_symlinks=False)
+                elif src.is_dir():
+                    children = src.iterdir()
+                    dst.mkdir(exist_ok=dirs_exist_ok)
```

Comment:
This feels a bit off, though it might be me misunderstanding. If `dirs_exist_ok` is false and the target exists, then the user will get a somewhat inscrutable error from a `.mkdir` call within a stack-based while loop. The first (and only) iteration will fail and no more work will happen. However, if `dirs_exist_ok` is true and the target exists, then many loops happen, joy is had, and many directories recurse off into the sunset. More importantly, for work to happen when the target exists in any way, `exist_ok` must be true.

If this understanding is correct, would it make sense to do check for `self.is_dir() and not dirs_exist_ok and target.exists()` before the copy-loop, and raise some sort of sensible error there? That would mean that you could then change this line to `dst.mkdir(exist_ok=True)`, which is what I expected to see here!

It is very possible I've gotten some or all of this wrong, apologies in advance if so.

Reply:
> If `dirs_exist_ok` is false and the target exists, then the user will get a somewhat inscrutable error from a `.mkdir` call within a stack-based while loop.

Hm. Is this traceback inscrutable? It seems reasonable and correct to me:

```python
>>> import pathlib
>>> pathlib.Path('Doc').copy('Lib')
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    pathlib.Path('Doc').copy('Lib')
    ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/home/barney/projects/cpython/Lib/pathlib/_abc.py", line 874, in copy
    on_error(err)
    ~~~~~~~~^^^^^
  File "/home/barney/projects/cpython/Lib/pathlib/_abc.py", line 852, in on_error
    raise err
  File "/home/barney/projects/cpython/Lib/pathlib/_abc.py", line 863, in copy
    dst.mkdir(exist_ok=dirs_exist_ok)
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barney/projects/cpython/Lib/pathlib/_local.py", line 774, in mkdir
    os.mkdir(self, mode)
    ~~~~~~~~^^^^^^^^^^^^
FileExistsError: [Errno 17] File exists: 'Lib'
```

Reply:
I did say a *somewhat* inscrutable error!

----

I agree it is understandable as it stands, but we could instead have:

```pycon
>>> import pathlib
>>> pathlib.Path('Doc').copy('Lib')
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    pathlib.Path('Doc').copy('Lib')
    ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/home/barney/projects/cpython/Lib/pathlib/_abc.py", line 000, in copy
    raise FileExistsError(f'Target directory exists: {target!r}')
    ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileExistsError: Target directory exists: 'Lib'
```

which cuts out three levels of traceback and has a slightly more descriptive error message.

Perhaps this is on the margins, but clarity of error messages is really useful, especially given that oftentimes errors occur in / create stressful environments!

A

Reply:
Gotcha!

It looks like we can eliminate a couple levels of traceback with a change to how `on_error` works - I'll do that now.

FWIW, [Serhiy](https://github.com/python/cpython/pull/118290#discussion_r1592907787) and [Eryk](https://github.com/python/cpython/pull/118290#discussion_r1619512183) pointed out to me that using "native" `OSError` produces exception messages in the user's local language, which is often desirable!

Reply:
Eliminating the `on_error` levels, we have:

```python
>>> pathlib.Path('Doc').copy('Lib')
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    pathlib.Path('Doc').copy('Lib')
    ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/home/barney/projects/cpython/Lib/pathlib/_abc.py", line 860, in copy
    dst.mkdir(exist_ok=dirs_exist_ok)
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barney/projects/cpython/Lib/pathlib/_local.py", line 774, in mkdir
    os.mkdir(self, mode)
    ~~~~~~~~^^^^^^^^^^^^
FileExistsError: [Errno 17] File exists: 'Lib'
```

Reply:
I'll resolve this thread because preferring 'native' `OSError`s is an established pattern in `pathlib` and related modules like `os.path`, partly because error messages are localised. Hope that's alright! Feel free to reopen if you'd like :)

Reply:
If it were me, I would probably add a pre-loop check & `raise OSError(errno.EEXIST)`. But that would add (some) complexity, and I think the change you made is a happy middle ground. Thanks!